applications/inkjet.md

```diff
@@ -0,0 +1,280 @@
+|       2. | Embedding Model | Replace default sentence embedding. Implement OpenAI's `text-embedding-ada-002` model, compatible with natural language and code processing. |
+|       3. | Vector Database | Update the `Milvus and Weaviate` vector database to house the code-based embeddings of the complete dataset. |
+
+
+### Milestone 3 — RAG-LLM Integration with Dataset, Efficacy Testing
+
+- **Estimated Duration:** 1 month
+- **FTE:**  3
+- **Costs:** 6,000 USD
+
+|  Number  | Deliverable | Specification |
+| -------: | ----------- | ------------- |
+|  **0a.** | License | Open-sourced under Apache 2.0. |
+|  **0b.** | Documentation | Code comments. Documentation provided for the integration process, and usage instructions provided for the updated RAG system. |
+|  **0c.** | Testing Guide | Unit tests updated to reflect changes in RAG system. Guide for testing the updated RAG system. |
+|  **0d.** | Docker | N/A |
+|       1. | RAG Integration | Integration of the new dataset and vector database with the prototype RAG system for improved retrieval. Modifiable parameters for queries, including temperature, top-k documents in retrieval, and token context size. |
+|       2. | Query Handling | Updated query parsing and processing using new embedding model to improve smart contract code retrieval. Prompt rewriting of plaintext in user request to improve semantic search. |
+|       3. | Testing Dataset | Create a dataset of prompts to test the output quality of RAG system, and if it generates vulnerabilities. |
+|       4. | System Testing | Testing of RAG retrieval functionality with the testing dataset. Evalaute performance and efficacy for vulnerability-free code. |
+
+### Milestone 4 — VS Code Extension Core Functionality
+
+- **Estimated Duration:** 1 month
+- **FTE:**  3
+- **Costs:** 6,000 USD
+
+|  Number  | Deliverable | Specification |
+| -------: | ----------- | ------------- |
+|  **0a.** | License | Open-sourced under Apache 2.0. |
+|  **0b.** | Documentation | Code comments. VS Code Extension UI components and functionality documented. Instructions for use documented. |
+|  **0c.** | Testing Guide | Guide on how to run tests on the VS Code extension. |
+|  **0d.** | Docker 1 | N/A |
+|       1. | UI/UX Design | Design and development of UI for VS Code extension. Extension will reside in primary sidebar, with multiple resizable sections. |
+|       2. | VS Code Extension | Creation of the core VS Code extension functionality, including integration with our RAG-LLM pipeline, code selection access, and local file system access. This includes Chat and Chat Settings functionality. Analysis and Templates functionality will be scaffolded (see Milestone 5). |
+|       3. | Usage Testing | Comprehensive testing to ensure the VS Code extension is responsive and stable. All parts of functionality will be tested. |
```

Comment:
We usually expect testing to be part of the development process of each component. Hence it's not necessarily to list it as a separate deliverable, since it's also difficult for us to verify.